### ISSUE TITLE
LZ4 compressing and decompressing of large messages doesn't work

### ISSUE THREAD

**Issue by graemerobertson** [Metaswitch/sas-client#68]
We created an LZ4Compressor and ran the following data (which is a Diameter message of length 4808 bytes, a number that is strictly bigger than 4096, which will become important later) through the compress function.

```
010012c8c000010f00000003d045efc632437f430000010740000044706572696d6574612d312e766d776172652e6f62752d656e6669656c642e746573743b313436333538333532333b35343637373b3139363630383035000001084000002a706572696d6574612d312e766d776172652e6f62752d656e6669656c642e7465737400000000012840000028706572696d6574612e766d776172652e6f62752d656e6669656c642e746573740000011b4000002762696c6c696e672e766d776172652e6f62752d656e6669656c642e7465737400000001034000000c00000003000001e04000000c00000002000001e54000000c00000000000000554000000c00000000000000014000002c2b313531303835383032373240766d776172652e6f62752d656e6669656c642e74657374000001164000000c573c8323000000374000000cdb4ca862000001cd40000018382e333232363040336770702e6f726700000369c0001128000028af0000036cc000111c000028af00000337c0000030000028af00000338c0000012000028af494e56495445000000000378c0000010000028af000000b40000033dc0000010000028af000000010000035ec0000010000028af000000010000033ec0000022000028af76356554525a6961576a67354f4747566f786336525100000000033fc0000036000028af3c7369703a2b313531303835383032373140766d776172652e6f62752d656e6669656c642e746573743e00000000033fc000001e000028af3c74656c3a2b31353130383538303237313e000000000340c000002f000028af7369703a2b3135313038353830323732403137322e31382e36332e3133333a353030300000000341c000004c000028af00000342c0000010000028afdb4ca86000000343c0000010000028afdb4ca862000008fdc0000010000028af000001b6000008fec0000010000028af000003a300000346c0000054000028af00000347c0000023000028af766d776172652e6f62752d656e6669656c642e746573740000000348c0000023000028af766d776172652e6f62752d656e6669656c642e746573740000000349c000002c000028af39313336626362393538653634623633386265626230306561383936356434380000034ac000000f000028af763d30000000034ac000003c000028af6f3d53414d53554e472d494d532d5545203132333435363130203120494e20495034203137322e31382e36332e3133330000034ac0000015000028af733d535320564f49500000000000034ac0000022000028af633d494e20495034203137322e31382e36332e31333300000000034ac0000013000028af623d41533a3431000000034ac0000014000028af623d52533a3531320000034ac0000015000028af623d52523a313533370000000000034ac0000011000028af743d3020300000000000034bc00003ec000028af0000034cc0000042000028af6d3d617564696f203137333038205254502f41565020313136203020382039362039203937203939203130302032203130312031313100000000034dc0000013000028af623d41533a3338000000034dc0000014000028af623d52533a3337350000034dc0000015000028af623d52523a313132350000000000034dc0000016000028af613d73656e647265637600000000034dc0000027000028af613d7274706d61703a31313620414d522d57422f31363030302f31000000034dc000002e000028af613d7274706d61703a3131312074656c6570686f6e652d6576656e742f313630303000000000034dc0000020000028af613d7274706d61703a302050434d552f383030300000034dc0000020000028af613d7274706d61703a382050434d412f383030300000034dc0000021000028af613d7274706d61703a39362053494c4b2f383030300000000000034dc0000020000028af613d7274706d61703a3920473732322f383030300000034dc0000020000028af613d7274706d61703a393720414d522f383030300000034dc0000024000028af613d7274706d61703a393920473732362d31362f383030300000034dc0000025000028af613d7274706d61703a31303020473732362d32342f383030300000000000034dc0000023000028af613d7274706d61703a3220473732362d33322f38303030000000034dc0000025000028af613d7274706d61703a31303120473732362d34302f383030300000000000034dc000003c000028af613d666d74703a313136206d6f64652d6368616e67652d6361706162696c6974793d323b206d61782d7265643d3232300000034dc000001b000028af613d666d74703a31313120302d3135000000034dc0000025000028af613d637572723a716f73206c6f63616c2073656e64726563760000000000034dc0000022000028af613d637572723a716f732072656d6f7465206e6f6e6500000000034dc000002e000028af613d6465733a716f73206d616e6461746f7279206c6f63616c2073656e647265637600000000034dc000002f000028af613d6465733a716f73206d616e6461746f72792072656d6f74652073656e6472656376000000034dc000001a000028af613d6d61787074696d653a32343000000000034dc0000023000028af613d666d74703a3937206f637465742d616c69676e3d31000000034dc0000016000028af613d7074696d653a3230000000000372c0000010000028af0000000100000508c0000034000028af7369703a2b313531303835383032373140766d776172652e6f62752d656e6669656c642e74657374000007f4c0000010000028af000000000000034bc0000274000028af0000034cc0000028000028af6d3d617564696f2031333134205254502f41565020313136203131310000034dc0000013000028af623d41533a3431000000034dc0000014000028af623d52533a3531320000034dc0000015000028af623d52523a313533370000000000034dc0000027000028af613d7274706d61703a31313620414d522d57422f31363030302f31000000034dc000003b000028af613d666d74703a313136206d6f64652d6368616e67652d6361706162696c6974793d323b6d61782d7265643d323230000000034dc000002e000028af613d7274706d61703a3131312074656c6570686f6e652d6576656e742f313630303000000000034dc000001b000028af613d666d74703a31313120302d3135000000034dc0000025000028af613d637572723a716f73206c6f63616c2073656e64726563760000000000034dc0000026000028af613d637572723a716f732072656d6f74652073656e647265637600000000034dc000002e000028af613d6465733a716f73206d616e6461746f7279206c6f63616c2073656e647265637600000000034dc000002f000028af613d6465733a716f73206d616e6461746f72792072656d6f74652073656e6472656376000000034dc0000016000028af613d73656e647265637600000000034dc0000016000028af613d7074696d653a323000000000034dc000001a000028af613d6d61787074696d653a323430000000000372c0000010000028af0000000100000508c0000034000028af7369703a2b313531303835383032373140766d776172652e6f62752d656e6669656c642e74657374000007f4c0000010000028af00000001000004f8c000080c000028af000004f9c000002c000028af000004fac0000010000028afdb4ca85f000004fbc0000010000028afdb4ca85f0000034bc0000540000028af0000034cc000004e000028af6d3d617564696f203137333038205254502f41565020313136203130372031313820393620302038203937203920393920313030203220313031203131312031313000000000034dc0000013000028af623d41533a3338000000034dc0000014000028af623d52533a3337350000034dc0000015000028af623d52523a313132350000000000034dc0000016000028af613d73656e646f6e6c7900000000034dc0000027000028af613d7274706d61703a31313620414d522d57422f31363030302f31000000034dc0000027000028af613d7274706d61703a31303720414d522d57422f31363030302f31000000034dc0000023000028af613d7274706d61703a31313820414d522f383030302f31000000034dc0000022000028af613d7274706d61703a393620414d522f383030302f3100000000034dc000002e000028af613d7274706d61703a3131312074656c6570686f6e652d6576656e742f313630303000000000034dc000002d000028af613d7274706d61703a3131302074656c6570686f6e652d6576656e742f383030300000000000034dc0000020000028af613d7274706d61703a302050434d552f383030300000034dc0000020000028af613d7274706d61703a382050434d412f383030300000034dc0000021000028af613d7274706d61703a39372053494c4b2f383030300000000000034dc0000020000028af613d7274706d61703a3920473732322f383030300000034dc0000024000028af613d7274706d61703a393920473732362d31362f383030300000034dc0000025000028af613d7274706d61703a31303020473732362d32342f383030300000000000034dc0000023000028af613d7274706d61703a3220473732362d33322f38303030000000034dc0000025000028af613d7274706d61703a31303120473732362d34302f383030300000000000034dc000003c000028af613d666d74703a313136206d6f64652d6368616e67652d6361706162696c6974793d323b206d61782d7265643d3232300000034dc000004a000028af613d666d74703a313037206f637465742d616c69676e3d313b6d6f64652d6368616e67652d6361706162696c6974793d323b206d61782d7265643d32323000000000034dc000003c000028af613d666d74703a313138206d6f64652d6368616e67652d6361706162696c6974793d323b206d61782d7265643d3232300000034dc0000049000028af613d666d74703a3936206f637465742d616c69676e3d313b6d6f64652d6368616e67652d6361706162696c6974793d323b206d61782d7265643d3232300000000000034dc000001b000028af613d666d74703a31313120302d3135000000034dc000001b000028af613d666d74703a31313020302d3135000000034dc0000021000028af613d637572723a716f73206c6f63616c206e6f6e650000000000034dc0000022000028af613d637572723a716f732072656d6f7465206e6f6e6500000000034dc000002e000028af613d6465733a716f73206d616e6461746f7279206c6f63616c2073656e647265637600000000034dc000002f000028af613d6465733a716f73206d616e6461746f72792072656d6f74652073656e6472656376000000034dc000001a000028af613d6d61787074696d653a32343000000000034dc0000016000028af613d7074696d653a3230000000000372c0000010000028af0000000100000508c0000034000028af7369703a2b313531303835383032373140766d776172652e6f62752d656e6669656c642e74657374000007f4c0000010000028af000000000000034bc0000294000028af0000034cc0000028000028af6d3d617564696f2031333134205254502f41565020313136203131310000034dc0000013000028af623d41533a3431000000034dc0000014000028af623d52533a3531320000034dc0000015000028af623d52523a313533370000000000034dc0000027000028af613d7274706d61703a31313620414d522d57422f31363030302f31000000034dc000003b000028af613d666d74703a313136206d6f64652d6368616e67652d6361706162696c6974793d323b6d61782d7265643d323230000000034dc000002e000028af613d7274706d61703a3131312074656c6570686f6e652d6576656e742f313630303000000000034dc000001b000028af613d666d74703a31313120302d3135000000034dc0000021000028af613d637572723a716f73206c6f63616c206e6f6e650000000000034dc0000022000028af613d637572723a716f732072656d6f7465206e6f6e6500000000034dc000002e000028af613d6465733a716f73206d616e6461746f7279206c6f63616c2073656e647265637600000000034dc000002f000028af613d6465733a716f73206d616e6461746f72792072656d6f74652073656e6472656376000000034dc0000026000028af613d636f6e663a716f732072656d6f74652073656e647265637600000000034dc0000016000028af613d73656e646f6e6c7900000000034dc0000016000028af613d7074696d653a323000000000034dc000001a000028af613d6d61787074696d653a323430000000000372c0000010000028af0000000100000508c0000034000028af7369703a2b313531303835383032373140766d776172652e6f62752d656e6669656c642e74657374000007f4c0000010000028af00000001000027108000004000004bf4000027118000003400004bf4000027128000001000004bf4636f7265000027138000001800004bf46163636573732d766f6c7465
```

We got the following output.

```
ff51010012c8c000010f00000003d045efc632437f430000010740000044706572696d6574612d312e766d776172652e6f62752d656e6669656c642e746573743b313436333538333532333b35343637373b3139363630383035000001084000002a44000fa4000000000128400000282c000f6e0005ff000000011b4000002762696c6c696e67270007f3000001034000000c00000003000001e00c005302000001e50c005700000000550c00ff03014000002c2b313531303835383032373240f20004f335000001164000000c573c8323000000374000000cdb4ca862000001cd40000018382e333232363040336770702e6f726700000369c0001128000028af0000036cc000111c0c005337c00000300c00f20a38c0000012000028af494e56495445000000000378c000001020006700b40000033d10005b010000035e1000f1103ec0000022000028af76356554525a6961576a67354f4747566f78633652515400e73fc0000036000028af3c7369703af1001f31f10005153e3800111e38003974656c38000220009b40c000002f000028af5700f20e32403137322e31382e36332e3133333a353030300000000341c000004cdc00240342c80087db4ca8600000034310005462000008fd100086000001b6000008fe1000a303a300000346c00000544c005147c000002304010fb102035f0000000348240010f12249c000002c000028af39313336626362393538653634623633386265626230306561383936356434380000034ac000000f7400113ddc00f91d4ac000003c000028af6f3d53414d53554e472d494d532d5545203132333435363130203120494e20495034201e01034c0011154801813d535320564f4950ac0201640001dc012f633d3a0003032400c413000028af623d41533a34311400121414006352533a353132140001640096623d52523a313533376400a211000028af743d3020301400534bc00003ec5c01f1304cc0000042000028af6d3d617564696f203137333038205254502f415650203131362030203820393620392039372039392031303020322031303120313131a400194da400243338140006a40033333735140007a400323132359000f5044dc0000016000028af613d73656e6472656376580012271800f1097274706d61703a31313620414d522d57422f31363030302f10015b4dc000002e2800f202312074656c6570686f6e652d6576656e7431000558001920300081302050434d552f38c9020f20000068382050434d412000192140007339362053494c4b21000400010a64006339204737323223000f64000068393720414d5220001a24640001410048362d3136240019252400b331303020473732362d32342500048c000140030548011232260014338f000244010c4c00123126002b34304c0001080352613d666d749201f4156d6f64652d6368616e67652d6361706162696c6974793d323b206d61782d7265643d3232f400121bb000033c006b3120302d31358000f800637572723a716f73206c6f63616c20130202a800013403072800b572656d6f7465206e6f6e65e001031002316465734b009f6d616e6461746f7279550000018c01122f98000e3000035d00078600013000121a3000c56d61787074696d653a3234307c00036c0101e400f5003937206f637465742d616c69676e3db00203f002033d00023c0016720405a1000100000508c000003438040fd7051447000007f4440003ec03270274ec0301b40605ec0339333134eb0301d20301e0000c74040ad0030274040bd0030574040fb80319123b4801012c010f4c020a094b0202640003e401051c040ff4030d0f7c02311226b00005a4020e1f02099c000e50020ed5020170010f80021d0340020797020278000318000958020fb002090f74023d9201000004f8c000080c74022404f9740744000004fa9c028adb4ca85f000004fb100062034bc000054038006f034cc000004e98060b613130372031319c0601a3063d372039a00636203131980306d4020fa4062e456f6e6c7958000fa406150b28001130f1050fcc060003400405d8021538190606f0060304050324001139170701a70501230002f4010f20031d122ddc020354003d31313044070221070460000ae0060f44072e1f374407210f2407c1124a88010214051b3041061f3b6e0714035804183c4c002f3138ac071a17493c002a3936c8060f870017028c020f7c05090a1c0019301408036809057005023f0504eb07025c00030c03052400039405042500016005030c030f9005460f1008090f9405010f78053d04ec075094000028af03d80b0fec07ff060f540295046c08316f6e66c50a03d8020ff407050d98060f7802010f0c085cf30327108000004000004bf400002711800000340c002112800811f00f4bf4636f7265000027138000001800004bf46163636573732d766f6c7465
```

We then changed https://github.com/Metaswitch/sas-client/blob/8456b35a741b61aa99fdcfb132fcc88c6826bfb6/source/sas_compress.cpp#L84 so that MAX_INPUT_SIZE was 8192 and did the compress again, getting a very similar output. This change caused us to only go through this for loop (https://github.com/Metaswitch/sas-client/blob/8456b35a741b61aa99fdcfb132fcc88c6826bfb6/source/sas_compress.cpp#L262) only once. The difference in the output is appears to be a few bytes towards the end of the message which are missing in this new output.

```
ff51010012c8c000010f00000003d045efc632437f430000010740000044706572696d6574612d312e766d776172652e6f62752d656e6669656c642e746573743b313436333538333532333b35343637373b3139363630383035000001084000002a44000fa4000000000128400000282c000f6e0005ff000000011b4000002762696c6c696e67270007f3000001034000000c00000003000001e00c005302000001e50c005700000000550c00ff03014000002c2b313531303835383032373240f20004f335000001164000000c573c8323000000374000000cdb4ca862000001cd40000018382e333232363040336770702e6f726700000369c0001128000028af0000036cc000111c0c005337c00000300c00f20a38c0000012000028af494e56495445000000000378c000001020006700b40000033d10005b010000035e1000f1103ec0000022000028af76356554525a6961576a67354f4747566f78633652515400e73fc0000036000028af3c7369703af1001f31f10005153e3800111e38003974656c38000220009b40c000002f000028af5700f20e32403137322e31382e36332e3133333a353030300000000341c000004cdc00240342c80087db4ca8600000034310005462000008fd100086000001b6000008fe1000a303a300000346c00000544c005147c000002304010fb102035f0000000348240010f12249c000002c000028af39313336626362393538653634623633386265626230306561383936356434380000034ac000000f7400113ddc00f91d4ac000003c000028af6f3d53414d53554e472d494d532d5545203132333435363130203120494e20495034201e01034c0011154801813d535320564f4950ac0201640001dc012f633d3a0003032400c413000028af623d41533a34311400121414006352533a353132140001640096623d52523a313533376400a211000028af743d3020301400534bc00003ec5c01f1304cc0000042000028af6d3d617564696f203137333038205254502f415650203131362030203820393620392039372039392031303020322031303120313131a400194da400243338140006a40033333735140007a400323132359000f5044dc0000016000028af613d73656e6472656376580012271800f1097274706d61703a31313620414d522d57422f31363030302f10015b4dc000002e2800f202312074656c6570686f6e652d6576656e7431000558001920300081302050434d552f38c9020f20000068382050434d412000192140007339362053494c4b21000400010a64006339204737323223000f64000068393720414d5220001a24640001410048362d3136240019252400b331303020473732362d32342500048c000140030548011232260014338f000244010c4c00123126002b34304c0001080352613d666d749201f4156d6f64652d6368616e67652d6361706162696c6974793d323b206d61782d7265643d3232f400121bb000033c006b3120302d31358000f800637572723a716f73206c6f63616c20130202a800013403072800b572656d6f7465206e6f6e65e001031002316465734b009f6d616e6461746f7279550000018c01122f98000e3000035d00078600013000121a3000c56d61787074696d653a3234307c00036c0101e400f5003937206f637465742d616c69676e3db00203f002033d00023c0016720405a1000100000508c000003438040fd7051447000007f4440003ec03270274ec0301b40605ec0339333134eb0301d20301e0000c74040ad0030274040bd0030574040fb80319123b4801012c010f4c020a094b0202640003e401051c040ff4030d0f7c02311226b00005a4020e1f02099c000e50020ed5020170010f80021d0340020797020278000318000958020fb002090f74023d9201000004f8c000080c74022404f9740744000004fa9c028adb4ca85f000004fb100062034bc000054038006f034cc000004e98060b613130372031319c0601a3063d372039a00636203131980306d4020fa4062e456f6e6c7958000fa406150b28001130f1050fcc060003400405d8021538190606f0060304050324001139170701a70501230002f4010f20031d122ddc020354003d31313044070221070460000ae0060f44072e1f374407210f2407c1124a88010214051b3041061f3b6e0714035804183c4c002f3138ac071a17493c002a3936c8060f870017028c020f7c05090a1c0019301408036809057005023f0504eb07025c00030c03052400039405042500016005030c030f9005460f1008090f9405010f78053d04ec07179440050fec07ff060f540295046c08316f6e66c50a03d8020ff407050d98060f7802010f0c085cf30327108000004000004bf400002711800000340c002112800811f00f4bf4636f7265000027138000001800004bf46163636573732d766f6c7465
```

Using the LZ4 ruby gem on SAS we were able to successfully decompress the second of these outputs, but not the first. We did this using (roughly) the following script.

```
require "lz4-ruby"

compressed = ["c825<BLAH>"].pack("H*")
output = LZ4::uncompress(compressed)
```

(c825 is the length of the decoded message encoded as the LZ4 gem expects)

We don't know whether the extra bytes in the compression when MAX_INPUT_SIZE was set to 4096 were erroneous, and therefore we have a problem in the compression side of things, or whether this is a decompression bug, or something else.

**Comment by graemerobertson** [comment]
@rkd-msw, you tried to write a UT which would reproduce this behaviour and it worked fine - any idea what might have been the key difference between your UT and what I've described above?

**Comment by rkd-msw** [comment]
Aha, got it - we're using `LZ4_compress_fast_continue` wrong. We're passing it two blocks of data to compress, and then joining the compressed output together into a single string, which we then try (and fail) to decompress.

Instead, we should be storing the compressed output of each block separately, and then calling LZ4_decompress_safe_continue on each block (passing it the same LZ4_streamDecode_t object each time to maintain the compression state). I've got some code where I do this and `LZ4_decompress_safe_continue` returns non-error values.

**Comment by rkd-msw** [comment]
(Or we should just be calling the single `LZ4_compress_fast`, and allocating a bigger buffer if it fails - this means we don't need any decompression-side changes.)

**Comment by rkd-msw** [comment]
Fixed in Clearwater by #69.
